numpy/core/shape_base.py

```diff
@@ -354,3 +359,163 @@ def stack(arrays, axis=0):
+    --------
+    concatenate : Join a sequence of arrays together.
+    stack : Stack arrays in sequence along a new dimension.
+    hstack : Stack arrays in sequence horizontally (column wise).
+    vstack : Stack arrays in sequence vertically (row wise).
+    dstack : Stack arrays in sequence depth wise (along third dimension).
+    vsplit : Split array into a list of multiple sub-arrays vertically.
+
+    Notes
+    -----
+    ``block`` is similar to Matlab's "square bracket stacking": ``[A A; B B]``
+
+    Examples
+    --------
+    Stacking scalars in a row:
+    >>> block([1, 2, 3])
+    array([1, 2, 3])
+
+    Stacking scalars with 1d arrays:
+    >>> a = np.array([2, 3])
+    >>> block([1, a])
+    np.array([1, 2, 3])
+
+    Stacking 1d arrays in a row:
+    >>> A = np.array([1, 2, 3])
+    >>> B = np.array([2, 3, 4])
+    >>> block([A, B])
+    array([1, 2, 3, 2, 3, 4])
+
+    Stacking 2d row-vectors in a row:
+    >>> A = np.array([[1, 2, 3]])
+    >>> B = np.array([[2, 3, 4]])
+    >>> block([A, B])
+    array([[1, 2, 3, 2, 3, 4]])
+
+    Stacking 1d arrays in a column:
+    >>> a = np.array([1, 1])
+    >>> b = np.array([2, 2])
+    >>> block([[a], [b]])
+    array([[1, 1],
+           [2, 2]])
+
+    Stacking 2d row-vectors in a column:
+    >>> A = np.array([[1, 2, 3]])
+    >>> B = np.array([[2, 3, 4]])
+    >>> block([[A], [B]])
+    array([[1, 2, 3],
+           [2, 3, 4]])
+
+    The tuple notation also works:
+    >>> A = np.ones((2, 2))
+    >>> B = 2 * A
+    >>> block((A, B))
+    array([[1, 1, 2, 2],
+           [1, 1, 2, 2]])
+
+    Block array with arbitrary shaped elements
+    >>> One = np.array([[1, 1, 1]])
+    >>> Two = np.array([[2, 2, 2]])
+    >>> Three = np.array([[3, 3, 3, 3, 3, 3]])
+    >>> four = np.array([4, 4, 4, 4, 4, 4])
+    >>> five = np.array([5])
+    >>> six = np.array([6, 6, 6, 6, 6])
+    >>> Zeros = np.zeros((2, 6), dtype=int)
+    >>> block([[One, Two],
+    ...         Three,
+    ...         four,
+    ...         [five, six],
+    ...         Zeros])
+    array([[1, 1, 1, 2, 2, 2],
+           [3, 3, 3, 3, 3, 3],
+           [4, 4, 4, 4, 4, 4],
+           [5, 6, 6, 6, 6, 6],
+           [0, 0, 0, 0, 0, 0],
+           [0, 0, 0, 0, 0, 0]])
+
+    """
+
+    def is_element(x):
+        return not isinstance(x, (list, tuple))
+
+    def recursive_map(x, base, aggregate=list):
+        """
+        Iterate over the nested list, applying `base` to items, and `aggregate`
+        to iterables of mapped items
+        """
+        def f(x):
+            if is_element(x):
+                return base(x)
+            else:
+                return aggregate(f(xi) for xi in x)
+        return f(x)
+
+    def exactly_nd(x, ndim):
+        x = asanyarray(x)
+        shape = [1] * ndim
+        shape[ndim-x.ndim:] = x.shape
+        return x.reshape(shape)
+
+    def max_or_0(xs):
+        """ Like max, but returns 0 on an empty iterable """
+        xs = list(xs)
+        return max(xs) if xs else 0
```

Comment:
While I know that most of the numpy devs prefer not to use `try-except` in their code, I think it has some value here because it prevents you from allocating an unnecessary list. Something like
```
try:
    return max(xs)
except ValueError:
    return 0
```